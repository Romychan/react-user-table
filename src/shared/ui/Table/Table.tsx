import { ReactNode, TableHTMLAttributes } from 'react';

import { cl } from '../../lib/utils';

import styles from './Table.module.scss';

interface ITableProps extends TableHTMLAttributes<HTMLTableElement> {
  /** Elements to display in the table: `<TableHeader />`, `<TableBody />` or other elements of the table */
  children: ReactNode;
}

/**
 * This component is used to display data in the form of a table.
 * Together with the `<TableHeader />`, `<TableBody />`, it allows to display and sort data.
 */
export const Table = ({ className, children, ...rest }: ITableProps) => {
  return (
    <table {...rest} className={cl(styles.table, className)}>
      {children}
    </table>
  );
};
