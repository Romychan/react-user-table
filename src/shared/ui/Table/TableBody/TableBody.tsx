import { cl } from '../../../lib/utils';
import { IColumnTable, MinTableItem } from '../types';
import { Loader } from '../../Loader';

import styles from './TableBody.module.scss';

/**
 * The interface of a table body component
 *
 * @template T Basic data interface
 * @template K List of keys from the `T` base interface
 */
interface ITableBodyProps<T, K extends keyof T> {
  /** The data for the table to display */
  data: T[];
  /** The list of columns to display in the table */
  columns: IColumnTable<T, K>[];
  /**
   * Whether the body of the table is loaded. If `true`, the `<Loader />` is displayed
   *
   * @default false
   */
  isLoading?: boolean;
  /**
   * Are there any errors in the request. If `true`, an empty table is displayed
   *
   * @default false
   */
  isError?: boolean;
  /**
   * The callback that will be called when you click on a row
   *
   * @param item Data from the selected row
   */
  onRowClick?: (item: T) => void;
}

/**
 * This is a component for displaying table data.
 *
 * @template T Basic data interface
 * @template K List of keys from the `T` base interface
 */
export const TableBody = <T extends MinTableItem, K extends keyof T>({
  data,
  columns,
  onRowClick,
  isLoading = false,
  isError = false,
}: ITableBodyProps<T, K>) => {
  const renderRows = () => {
    if ((!isLoading && !data.length) || isError) {
      return (
        <tr className={styles.placeholder}>
          <td colSpan={columns.length} data-testid="table-body-empty">
            Empty Data
          </td>
        </tr>
      );
    } else if (data.length) {
      return data.map((item, index) => {
        return (
          <tr
            key={item.id}
            className={cl(styles.row, { [styles.clickable]: !!onRowClick })}
            aria-rowindex={index}
            onClick={() => (onRowClick ? onRowClick(item) : undefined)}
            data-testid="table-body-row"
          >
            {columns.map(({ accessor, renderCell }) => (
              <td
                key={accessor as string}
                className={styles.cell}
                data-testid={`table-body-cell-${accessor as string}`}
              >
                {renderCell
                  ? renderCell({ currentData: item })
                  : (item[accessor] as string)}
              </td>
            ))}
          </tr>
        );
      });
    } else {
      return (
        <tr className={styles.placeholder}>
          <td colSpan={columns.length}>
            <Loader />
          </td>
        </tr>
      );
    }
  };

  return (
    <tbody data-testid="table-body" className={styles.body}>
      {renderRows()}
    </tbody>
  );
};
