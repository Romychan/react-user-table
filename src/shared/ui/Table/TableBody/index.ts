export { TableBody } from './TableBody';
export {
  Default as TableBodyDefault,
  Loading as TableBodyLoading,
  Empty as TableBodyEmpty,
} from './TableBody.stories';
