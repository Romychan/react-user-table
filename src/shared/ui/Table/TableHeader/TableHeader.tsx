import { TableOrderValue, IColumnTable, ISortingTable } from '../types';
import { TableHeaderCell } from '../TableHeaderCell';

import styles from './TableHeader.module.scss';

/**
 * The interface of a table header component
 *
 * @template T Basic data interface
 * @template K List of keys from the `T` base interface
 */
interface ITableHeaderProps<T, K extends keyof T> {
  /** The list of columns to display in the table */
  columns: IColumnTable<T, K>[];
  /** Column name and sort type for sorting values */
  sorting: ISortingTable<T, K>;
  /**
   * The callback that will be called when sorting is selected
   *
   * @param accessor The key of the current sorting column
   * @param order The type of current sorting
   */
  onSort?: (accessor: K, order: TableOrderValue) => void;
}

/**
 * This is a component for displaying table column headers
 *
 * @template T Basic data interface
 * @template K List of keys from the `T` base interface
 */
export const TableHeader = <T, K extends keyof T>({
  columns,
  sorting,
  onSort,
}: ITableHeaderProps<T, K>) => {
  const handleSortChange = (accessor: K) => {
    if (onSort) {
      const sortOrder =
        sorting.column === accessor && sorting.order === 'ascending'
          ? 'descending'
          : 'ascending';

      onSort(accessor, sortOrder);
    }
  };

  return (
    <thead className={styles.header}>
      <tr className={styles.row}>
        {columns.map(({ label, accessor, width, isSortable, isResizable }) => (
          <TableHeaderCell
            key={accessor as string}
            order={sorting.order}
            width={width}
            isSortable={isSortable}
            isActive={accessor === sorting.column}
            isResizable={isResizable}
            onClick={() => handleSortChange(accessor)}
          >
            {label}
          </TableHeaderCell>
        ))}
      </tr>
    </thead>
  );
};
