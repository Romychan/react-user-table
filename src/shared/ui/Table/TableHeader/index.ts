export { TableHeader } from './TableHeader';
export {
  Default as TableHeaderDefault,
  Resizable as TableHeaderResizable,
} from './TableHeader.stories';
