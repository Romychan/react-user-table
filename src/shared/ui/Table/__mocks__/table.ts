import { IColumnTable, ISortingTable } from '..';

export interface MockUser {
  id: number;
  username: string;
  name: string;
  email: string;
}

export const MOCK_COLUMNS: IColumnTable<MockUser, keyof MockUser>[] = [
  { label: 'ID', accessor: 'id', width: 120, isSortable: true },
  {
    label: 'Profile',
    accessor: 'name',
    width: 240,
    isSortable: true,
    renderCell: ({ currentData }) =>
      `${currentData.name} ${currentData.username}`,
  },
  { label: 'Email', accessor: 'email', width: 240 },
];

export const MOCK_SORTING: ISortingTable<MockUser, keyof MockUser> = {
  column: 'id',
  order: 'ascending',
};

export const MOCK_DATA_TABLE: MockUser[] = [
  {
    id: 1,
    name: 'Test Name',
    username: 'testname',
    email: 'test@test.test',
  },
];
