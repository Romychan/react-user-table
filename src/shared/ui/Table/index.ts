export { Table } from './Table';
export { TableBody } from './TableBody';
export { TableHeader } from './TableHeader';

export type {
  TableOrderValue,
  MinTableItem,
  ISortingTable,
  IRenderedCell,
  IColumnTable,
} from './types';
