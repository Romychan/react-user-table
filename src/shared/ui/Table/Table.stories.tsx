import type { Meta, StoryObj } from '@storybook/react';

import { withWidthDecorator } from '../../lib/storybook';

import { MOCK_COLUMNS, MOCK_SORTING } from './__mocks__/table';
import { Table } from './Table';
import { TableBody, TableHeader } from '.';
import {
  TableBodyDefault,
  TableBodyEmpty,
  TableBodyLoading,
} from './TableBody';
import { TableHeaderDefault, TableHeaderResizable } from './TableHeader';

const meta: Meta<typeof Table> = {
  title: 'UI/Table',
  component: Table,
  parameters: {
    layout: 'centered',
    width: '600px',
  },
  decorators: [withWidthDecorator],
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

/** This state is used to default. */
export const Default: Story = {
  render: function Render(args) {
    return (
      <Table {...args}>
        <TableHeader
          {...TableHeaderDefault.args}
          columns={TableHeaderDefault.args?.columns || []}
          sorting={TableHeaderDefault.args?.sorting || MOCK_SORTING}
        />
        <TableBody
          {...TableBodyDefault.args}
          columns={TableBodyDefault.args?.columns || []}
          data={TableBodyDefault.args?.data || []}
        />
      </Table>
    );
  },
};

/** To resize a cell, use the `isResizable` property for a column in the columns list. */
export const Resizable: Story = {
  args: {
    style: {
      width: 'max-content',
    },
  },
  render: function Render(args) {
    return (
      <div style={{ overflowX: 'auto' }}>
        <Table {...args}>
          <TableHeader
            {...TableHeaderResizable.args}
            columns={TableHeaderResizable.args?.columns || []}
            sorting={TableHeaderResizable.args?.sorting || MOCK_SORTING}
          />
          <TableBody
            {...TableBodyDefault.args}
            columns={TableBodyDefault.args?.columns || []}
            data={TableBodyDefault.args?.data || []}
          />
        </Table>
      </div>
    );
  },
};

/**
 * The property `isLoading` indicates whether the table body is loaded.
 * If the value is `true`, the `<Loader />` is displayed.
 */
export const Loading: Story = {
  render: function Render(args) {
    return (
      <Table {...args}>
        <TableHeader
          {...TableHeaderDefault.args}
          columns={TableHeaderDefault.args?.columns || []}
          sorting={TableHeaderDefault.args?.sorting || MOCK_SORTING}
        />
        <TableBody
          {...TableBodyLoading.args}
          columns={TableBodyLoading.args?.columns || MOCK_COLUMNS}
          data={[]}
        />
      </Table>
    );
  },
};

/** If the data list is empty and the load is complete, an empty table will be displayed */
export const Empty: Story = {
  render: function Render(args) {
    return (
      <Table {...args}>
        <TableHeader
          {...TableHeaderDefault.args}
          columns={TableHeaderDefault.args!.columns || []}
          sorting={TableHeaderDefault.args?.sorting || MOCK_SORTING}
        />
        <TableBody
          {...TableBodyEmpty.args}
          columns={TableBodyEmpty.args?.columns || MOCK_COLUMNS}
          data={[]}
        />
      </Table>
    );
  },
};
