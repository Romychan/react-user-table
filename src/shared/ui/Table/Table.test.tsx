import { render, screen, RenderResult } from '@testing-library/react';

import { MOCK_COLUMNS, MOCK_DATA_TABLE } from './__mocks__/table';
import { Table } from './Table';
import { TableBody } from './TableBody';
import { TableHeader } from './TableHeader';

describe('Table', () => {
  let container: RenderResult;

  beforeEach(() => {
    container = render(
      <Table>
        <TableHeader
          columns={MOCK_COLUMNS}
          sorting={{ column: 'id', order: 'ascending' }}
        />
        <TableBody columns={MOCK_COLUMNS} data={MOCK_DATA_TABLE} />
      </Table>,
    );
  });

  it('should render correctly', () => {
    const headerCellElement = screen.getAllByTestId('table-header-cell');
    const tableBodyElement = screen.getByTestId('table-body');

    expect(headerCellElement).toHaveLength(3);
    expect(tableBodyElement).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
