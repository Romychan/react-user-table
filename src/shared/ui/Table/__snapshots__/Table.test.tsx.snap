// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="_table_78892f"
      >
        <thead
          class="_header_5dd620"
        >
          <tr
            class="_row_5dd620"
          >
            <th
              aria-sort="ascending"
              class="_cell_de2dbc _active_de2dbc _isSortable_de2dbc _asc_de2dbc"
              data-testid="table-header-cell"
              style="width: 120px;"
            >
              ID
              <svg
                class="icon icon-chevron _icon_de2dbc"
                data-testid="icon"
                height="14"
                width="14"
              >
                <use
                  xlink:href="/images/icons.svg#chevron"
                />
              </svg>
              <div
                class="_resizer_de2dbc"
                data-testid="table-header-cell-resizer"
              />
            </th>
            <th
              aria-sort="none"
              class="_cell_de2dbc _isSortable_de2dbc"
              data-testid="table-header-cell"
              style="width: 240px;"
            >
              Profile
              <svg
                class="icon icon-chevron _icon_de2dbc"
                data-testid="icon"
                height="14"
                width="14"
              >
                <use
                  xlink:href="/images/icons.svg#chevron"
                />
              </svg>
              <div
                class="_resizer_de2dbc"
                data-testid="table-header-cell-resizer"
              />
            </th>
            <th
              aria-sort="none"
              class="_cell_de2dbc"
              data-testid="table-header-cell"
              style="width: 240px;"
            >
              Email
              <div
                class="_resizer_de2dbc"
                data-testid="table-header-cell-resizer"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="_body_f9eda9"
          data-testid="table-body"
        >
          <tr
            aria-rowindex="0"
            class="_row_f9eda9"
            data-testid="table-body-row"
          >
            <td
              class="_cell_f9eda9"
              data-testid="table-body-cell-id"
            >
              1
            </td>
            <td
              class="_cell_f9eda9"
              data-testid="table-body-cell-name"
            >
              Test Name testname
            </td>
            <td
              class="_cell_f9eda9"
              data-testid="table-body-cell-email"
            >
              test@test.test
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="_table_78892f"
    >
      <thead
        class="_header_5dd620"
      >
        <tr
          class="_row_5dd620"
        >
          <th
            aria-sort="ascending"
            class="_cell_de2dbc _active_de2dbc _isSortable_de2dbc _asc_de2dbc"
            data-testid="table-header-cell"
            style="width: 120px;"
          >
            ID
            <svg
              class="icon icon-chevron _icon_de2dbc"
              data-testid="icon"
              height="14"
              width="14"
            >
              <use
                xlink:href="/images/icons.svg#chevron"
              />
            </svg>
            <div
              class="_resizer_de2dbc"
              data-testid="table-header-cell-resizer"
            />
          </th>
          <th
            aria-sort="none"
            class="_cell_de2dbc _isSortable_de2dbc"
            data-testid="table-header-cell"
            style="width: 240px;"
          >
            Profile
            <svg
              class="icon icon-chevron _icon_de2dbc"
              data-testid="icon"
              height="14"
              width="14"
            >
              <use
                xlink:href="/images/icons.svg#chevron"
              />
            </svg>
            <div
              class="_resizer_de2dbc"
              data-testid="table-header-cell-resizer"
            />
          </th>
          <th
            aria-sort="none"
            class="_cell_de2dbc"
            data-testid="table-header-cell"
            style="width: 240px;"
          >
            Email
            <div
              class="_resizer_de2dbc"
              data-testid="table-header-cell-resizer"
            />
          </th>
        </tr>
      </thead>
      <tbody
        class="_body_f9eda9"
        data-testid="table-body"
      >
        <tr
          aria-rowindex="0"
          class="_row_f9eda9"
          data-testid="table-body-row"
        >
          <td
            class="_cell_f9eda9"
            data-testid="table-body-cell-id"
          >
            1
          </td>
          <td
            class="_cell_f9eda9"
            data-testid="table-body-cell-name"
          >
            Test Name testname
          </td>
          <td
            class="_cell_f9eda9"
            data-testid="table-body-cell-email"
          >
            test@test.test
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
